File: RAG/query-rag.ts
#!/usr/bin/env node

import { ChromaClient } from 'chromadb';
import { GoogleGenAI } from '@google/genai';
import { config } from 'dotenv';

config();

class RAGQuerySystem {
  private client: any;
  private collection: any;
  private genAI: GoogleGenAI;

  constructor() {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
      throw new Error("GEMINI_API_KEY not found in environment variables");
    }
    this.genAI = new GoogleGenAI({apiKey});
    this.setupComponents();
  }

  private async setupComponents() {
    this.client = new ChromaClient();
    this.collection = await this.client.getCollection({ name: "rag_documents" });
    console.log("✅ Connected to vector database");
    console.log("✅ Connected to Gemini API");
  }

  private async retrieveRelevantContext(query: string, nResults: number = 3): Promise<string> {
    const result = await this.genAI.models.embedContent({model: "text-embedding-004", contents: [query]});
    const queryEmbedding = result.embeddings?.[0]?.values;

    const results = await this.collection.query({
      queryEmbeddings: [queryEmbedding],
      nResults
    });

    const documents = results.documents?.[0] || [];
    return documents.join("\n\n---\n\n");
  }

  private async generateAnswer(query: string, context: string): Promise<string> {
    const prompt = `Based on the following context, please answer the question. If the context doesn't contain relevant information to answer the question, please say so.

Context:
${context}

Question: ${query}

Answer:`;

    const result = await this.genAI.models.generateContent({model: "gemini-1.5-flash", contents: [prompt]});
    
    return result.candidates?.[0]?.content?.parts?.[0]?.text || "no answer";
  }

  async query(question: string) {
    console.log(`🔍 Query: ${question}`);
    console.log("\n" + "=".repeat(50));

    console.log("📚 Retrieving relevant context...");
    const context = await this.retrieveRelevantContext(question);

    console.log(`📄 Retrieved Context:`);
    console.log("-".repeat(30));
    console.log(context);
    console.log("-".repeat(30));

    console.log("\n🤖 Generating answer with Gemini...");
    const answer = await this.generateAnswer(question, context);

    console.log(`\n💬 Answer:`);
    console.log("=".repeat(50));
    console.log(answer);
    console.log("=".repeat(50));
  }
}

async function main() {
  const question = process.argv[2] || "what does kriti long for";
  const rag = new RAGQuerySystem();
  
  setTimeout(async () => {
    await rag.query(question);
  }, 2000);
}

main().catch(console.error);

File: RAG-python/query_rag.py
#!/usr/bin/env python3
"""
Script to query the RAG system using command line input and Gemini API.
Usage: python query_rag.py "your question here"
"""

import os
import sys
import argparse
import chromadb
from sentence_transformers import SentenceTransformer
import google.generativeai as gemini
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class RAGQuerySystem:
    def __init__(self):
        self.client = None
        self.collection = None
        self.embedding_model = None
        self.gemini_model = None
        self.setup_components()
    
    def setup_components(self):
        """Initialize all components of the RAG system."""
        # Setup ChromaDB
        self.client = chromadb.PersistentClient(path="./chroma_db")
        self.collection = self.client.get_collection("rag_documents")
        print("✅ Connected to vector database")
        
        # Setup embedding model (same as used for creation)
        self.embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
        print("✅ Loaded embedding model")
        
        # Setup Gemini API
        api_key = os.getenv('GEMINI_API_KEY')
        
        gemini.configure(api_key=api_key)
        self.gemini_model = gemini.GenerativeModel('gemini-1.5-flash')
        print("✅ Connected to Gemini API")
    
    def retrieve_relevant_context(self, query: str, n_results: int = 3) -> str:
        """Retrieve relevant context from vector database."""
        # Create embedding for the query
        query_embedding = self.embedding_model.encode([query])
        
        # Search in ChromaDB
        results = self.collection.query(
            query_embeddings=query_embedding.tolist(),
            n_results=n_results
        )
        
        # Extract and combine relevant documents
        documents = results['documents'][0] if results['documents'] else []
        
        # Combine documents with separators
        context = "\n\n---\n\n".join(documents)
        return context
        

    
    def generate_answer(self, query: str, context: str) -> str:
        """Generate answer using Gemini API with retrieved context."""
        prompt = f"""Based on the following context, please answer the question. If the context doesn't contain relevant information to answer the question, please say so.

        Context:
        {context}

        Question: {query}

        Answer:"""
                    
        response = self.gemini_model.generate_content(prompt)
        return response.text

    
    def query(self, question: str):
        """Main query method that retrieves context and generates answer."""
        print(f"🔍 Query: {question}")
        print("\n" + "="*50)
        
        # Retrieve relevant context
        print("📚 Retrieving relevant context...")
        context = self.retrieve_relevant_context(question)
        
        print(f"📄 Retrieved Context:")
        print("-" * 30)
        print(context)
        print("-" * 30)
        
        # Generate answer using Gemini
        print("\n🤖 Generating answer with Gemini...")
        answer = self.generate_answer(question, context)
        
        print(f"\n💬 Answer:")
        print("=" * 50)
        print(answer)
        print("=" * 50)

def main():
    rag = RAGQuerySystem()
    
    # Process query
    rag.query("What is the moon made of")

if __name__ == "__main__":
    main() 